rbd/migration.go

```diff
@@ -0,0 +1,240 @@
+	ret := C.rbd_migration_commit(
+		cephIoctx(ioctx),
+		cName)
+	return getError(ret)
+}
+
+// MigrationAbort aborts a migration in progress
+// breaking the relationship of image to the source
+//
+// Implements:
+//  int rbd_migration_abort(rados_ioctx_t ioctx,
+// 							const char *image_name);
+//
+func MigrationAbort(ioctx *rados.IOContext, name string) error {
+	cName := (*C.char)(C.CBytes([]byte(name)))
+
+	defer func() {
+		C.free(unsafe.Pointer(cName))
+	}()
+
+	ret := C.rbd_migration_abort(
+		cephIoctx(ioctx),
+		cName)
+	return getError(ret)
+}
+
+// MigrationImageState denotes the current migration status of a given image
+type MigrationImageState int
+
+const (
+	// MigrationImageUnknown is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_UNKNOWN from librbd.
+	MigrationImageUnknown = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_UNKNOWN)
+	// MigrationImageError is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ERROR from librbd.
+	MigrationImageError = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ERROR)
+	// MigrationImagePreparing is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARING from librbd.
+	MigrationImagePreparing = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARING)
+	// MigrationImagePrepared is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARED from librbd.
+	MigrationImagePrepared = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARED)
+	// MigrationImageExecuting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTING from librbd.
+	MigrationImageExecuting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTING)
+	// MigrationImageExecuted is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTED from librbd.
+	MigrationImageExecuted = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTED)
+	// MigrationImageAborting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ABORTING from librbd.
+	MigrationImageAborting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ABORTING)
+)
+
+// MigrationImageInfo provides information about the
+// live migration progress of an image
+type MigrationImageInfo struct {
+	SourcePoolID        int
+	SourcePoolNamespace string
+	SourceImageName     string
+	SourceImageID       string
+	DestPoolID          int
+	DestPoolNamespace   string
+	DestImageName       string
+	DestImageID         string
+	State               MigrationImageState
+	StateDescription    string
+}
+
+// MigrationStatus retrieve status of a live migration
+// for the specified image
+//
+// Implements:
+//  int rbd_migration_status(rados_ioctx_t ioctx,
+// 							 const char *image_name,
+// 							 rbd_image_migration_status_t *status,
+// 							 size_t status_size);
+//
+func MigrationStatus(ioctx *rados.IOContext, name string) (*MigrationImageInfo, error) {
+	cName := (*C.char)(C.CBytes([]byte(name)))
+
+	defer func() {
+		C.free(unsafe.Pointer(cName))
+	}()
+
+	var status C.rbd_image_migration_status_t
+	ret := C.rbd_migration_status(
```

Comment:
I think that, on success, this function call will allocate memory for the values stored in the                                       rbd_image_migration_status_t. The librbd header provides a function `rbd_migration_status_cleanup` to free this memory. We need to call this function after converting the C types to Go types in order to free the memory that the `rbd_migration_status` call allocated.

Reply:
FWIW, you can double check the ceph code to see if `rbd_migration_status_cleanup` is a noop when `rbd_image_migration_status_t*` or the fields are null and if so implementing the cleanup can be done with a simple defer.